src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonNode.Traversal.cs

```diff
@@ -195,13 +197,13 @@ void AddNewPair(JsonNode jsonNode, bool keepInCurrentNodes = false)
                         AddToParent(currentPair, ref currentNodes, ref toReturn, options.DuplicatePropertyNameHandling);
                         break;
                     case JsonTokenType.PropertyName:
-                        currentNodes.Push(new KeyValuePair<string, JsonNode>(reader.GetString(), null));
+                        currentNodes.Push(new KeyValuePair<string?, JsonNode?>(reader.GetString(), null));
                         break;
                     case JsonTokenType.Number:
                         AddNewPair(new JsonNumber(JsonHelpers.Utf8GetString(reader.ValueSpan)));
                         break;
                     case JsonTokenType.String:
-                        AddNewPair(new JsonString(reader.GetString()));
+                        AddNewPair(new JsonString(reader.GetString()!));
```

Comment:
`reader.GetString()` returns null only when its `TokenType` is `Null`. Here token type is `String` so will never be null.

Above `DeepCopy` it is `JsonElement.GetString()` which can be null if `row.TokenType is Null` about which I was not sure what will be that value, most likely it will not be `JsonTokenType.Null` but cannot guarantee

I other words here it checks TokenType of same instance which is `String`, but above not sure about the flow, not sure what/how those rows are populated and what its TokenType will be
```csharp
    DbRow row = _parsedData.Get(index);

    JsonTokenType tokenType = row.TokenType;

    if (tokenType == JsonTokenType.Null)
    {
        return null;
    }
```

Reply:
> I other words here it checks TokenType of same instance which is `String`, but above not sure about the flow, not sure what/how those rows are populated and what its TokenType will be

Effectively, the same thing is happening in both cases. If the JsonElement points to `null`, we return null (and the `ValueKind is null). If the reader points to a token that is `null`, we return null (and the TokenType is null).

If you look at the following methods implementation (`Get(...)` and `row.TokenType` property), it is effectively identical to what `ValueKind` property does:
```C#
DbRow row = _parsedData.Get(index);

JsonTokenType tokenType = row.TokenType;
```

**That said**, I can see why the code flow is more complex here, and it makes sense to `Debug.Assert` in one and `!` in the other. Thanks for the explanation.